### DISCUSSION TITLE
cli hyperlinks with `style_hyperlink()` no longer seem to work

### DISCUSSION THREAD

**Discussion by olivroy** [posit-dev/positron#4968] [Q&A] [answered]
```r
cli::style_hyperlink(
  "file","file://NEWS.md", list(line = 6, col = 2)
)
```

Try to click on this link generated from this. It works in RStudio. (with NEWS.md being an existing file)

On Windows, it opens the following (non-existing) file
![image](https://github.com/user-attachments/assets/7c5d6899-f4c9-4966-9a8c-cfc2383f0bf2)

Edit: changed to 2 leading slashes as this is what is *actually* supported in RStudio.

Edit2: 

This seems to work.

```r
cli::style_hyperlink(
  "file","file://NEWS.md", c(line = 6, col = 2)
)
```

**Comment by jennybc** [comment]
I am also seeing this (file hyperlinks no longer working) in:

Positron Version: 2024.10.0 (Universal) build 14
Code - OSS Version: 1.93.0
Commit: 901ab5d11d694212d32e53b97f771c5d601e428e
Date: 2024-10-09T15:26:55.363Z
Electron: 30.4.0
Chromium: 124.0.6367.243
Node.js: 20.15.1
V8: 12.4.254.20-electron.0
OS: Darwin arm64 22.6.0

**Comment by jennybc** [comment]
I'm seeing this in developer console:

```
ERR [UriError]: If a URI does not contain an authority component, then the path cannot begin with two slash characters ("//"): Error: [UriError]: If a URI does not contain an authority component, then the path cannot begin with two slash characters ("//")
    at _validateUri (vscode-file://vscode-app/Users/jenny/rrr/positron/out/vs/base/common/uri.js:61:27)
    at new URI (vscode-file://vscode-app/Users/jenny/rrr/positron/out/vs/base/common/uri.js:152:17)
    at new Uri (vscode-file://vscode-app/Users/jenny/rrr/positron/out/vs/base/common/uri.js:371:13)
    at URI.parse (vscode-file://vscode-app/Users/jenny/rrr/positron/out/vs/base/common/uri.js:243:20)
    at OpenerService.open (vscode-file://vscode-app/Users/jenny/rrr/positron/out/vs/editor/browser/services/openerService.js:174:70)
    at hyperlinkClickHandler (vscode-file://vscode-app/Users/jenny/rrr/positron/out/vs/workbench/browser/positronAnsiRenderer/outputRun.js:131:37)
    at Object.lj (vscode-file://vscode-app/Users/jenny/rrr/positron/node_modules/react-dom/umd/react-dom.production.min.js:223:217)
    at kj (vscode-file://vscode-app/Users/jenny/rrr/positron/node_modules/react-dom/umd/react-dom.production.min.js:34:117)
    at nj (vscode-file://vscode-app/Users/jenny/rrr/positron/node_modules/react-dom/umd/react-dom.production.min.js:34:171)
    at gh (vscode-file://vscode-app/Users/jenny/rrr/positron/node_modules/react-dom/umd/react-dom.production.min.js:62:93)
    at Xg (vscode-file://vscode-app/Users/jenny/rrr/positron/node_modules/react-dom/umd/react-dom.production.min.js:63:1)
    at vscode-file://vscode-app/Users/jenny/rrr/positron/node_modules/react-dom/umd/react-dom.production.min.js:72:332
    at Tf (vscode-file://vscode-app/Users/jenny/rrr/positron/node_modules/react-dom/umd/react-dom.production.min.js:189:448)
    at wg (vscode-file://vscode-app/Users/jenny/rrr/positron/node_modules/react-dom/umd/react-dom.production.min.js:32:481)
    at Ce (vscode-file://vscode-app/Users/jenny/rrr/positron/node_modules/react-dom/umd/react-dom.production.min.js:65:218)
    at Be (vscode-file://vscode-app/Users/jenny/rrr/positron/node_modules/react-dom/umd/react-dom.production.min.js:47:62)
    at Aj (vscode-file://vscode-app/Users/jenny/rrr/positron/node_modules/react-dom/umd/react-dom.production.min.js:46:351)
```

**Reply by olivroy** [reply]
Thanks for looking into this. FWIW, I edited my post. The supported syntax is with 2 leading `/`, not 4. but same result.

Do you think this bug is a Positron or cli bug?

**Comment by jennybc** [comment]
This commit looks likely to be the culprit (but this should be taken with a big grain of salt):

https://github.com/posit-dev/positron/commit/833a9b33b02188ef9d7e3f057064d537cbb77d9d#diff-b68e95d8ebaec9fc38aad38c3df9f68d2951bc577b3de330bf552bcafa7f1d34R619

**Comment by jennybc** [comment]
I think it's a positron bug.

**Comment by jennybc** [comment]
By the time there's an attempt to open the file, it has become `//NEWS.md/`

<img width="1238" alt="Screenshot 2024-10-09 at 1 23 09 PM" src="https://github.com/user-attachments/assets/c6da7603-4341-45e7-bf5f-52b26e0c06bb">

Screenshot is with

Positron Version: 2024.10.0 (Universal) build 14
Code - OSS Version: 1.93.0
Commit: 901ab5d11d694212d32e53b97f771c5d601e428e
Date: 2024-10-09T15:26:55.363Z
Electron: 30.4.0
Chromium: 124.0.6367.243
Node.js: 20.15.1
V8: 12.4.254.20-electron.0
OS: Darwin arm64 22.6.0

**Comment by jennybc** [comment]
I am _not_ seeing a problem with this clickable link:

```r
cli::cli_text("{.file NEWS.md}")
```

so that suggests there may be an interaction between positron changes and this style of file url `file://NEWS.md` specifically.

**Comment by jennybc** [comment]
Here's the working file hyperlink, produce via inline markup:

> cli::format_inline("{.file NEWS.md}")
[1] "\033]8;;file:///Users/jenny/rrr/usethis/NEWS.md\a\033[34mNEWS.md\033[39m\033]8;;\a"

So it's using an absolute path that starts with `file:///`.

**Reply by jennybc** [reply]
Based on what I see here in cli's internal docs, I'm not sure `file://NEWS.md` was ever meant to work 🤔 maybe that was working by coincidence and changes in Positron have broken it, but it's not really a bug? Because file hyperlinks made with cli's official inline markup do appear to be working.

https://github.com/r-lib/cli/blob/main/R/ansi-hyperlink.R

**Reply by olivroy** [reply]
Due to https://github.com/r-lib/cli/issues/601, I have used this syntax as a workaround. It is used in roxygen2's code to generate warnings   with clickable links at file line numbers + different display names. I have sprinkled this in my own code and packages I contribute to like lintr and goodpractice (maybe pkgdown too) https://cli.r-lib.org/reference/links.html?q=.href#files

**Reply by olivroy** [reply]
After a bit more investigation, I changed my internal code to using 

```r
cli::style_hyperlink(
  "file","file://NEWS.md", c(line = 6, col = 2)
)
```

like it is in roxygen2 in `warn_roxy()` and I am seeing no problems so far. https://github.com/olivroy/reuseme/commit/fe46ea277058d56c49fa47bf02daafb1a97a5ed8

https://github.com/r-lib/roxygen2/blob/9652d15221109917d46768e836eaf55e33c21633/R/utils-warn.R#L24-L34

I simulated a roxygen2 warning (i.e. add an empty `#' @examples` block to a function and called `devtools::document()` and to my surprise, the clickable link worked.

**Reply by jennybc** [reply]
So is this an accurate summary of where we are?

1. The original example was using a suboptimal call to `cli::style_hyperlink()` that happened to work in older Positron, but doesn't work in the most recent release.
2. If you update to a calling pattern you see in comparable use cases elsewhere, the hyperlinks work in current Positron.

If this is the case, then I guess there's no follow up needed (?).

**Reply by olivroy** [reply]
I think this is right. 

From a github search for `style_hyperlink()`, I found 2 usages of `params = list()`

https://github.com/rossellhayes/wtfn/blob/f223d4926e8590b043181b9c103a2fdaaf1dace1/R/get_wtfn_status.R#L133
https://github.com/JorisChau/checkglobals/blob/f0d830907e41529bff8aceffc12cea22020202a3/R/print.R#L318

Is there a way to add a test in Positron to make sure the correct syntax (with params as vector) doesn't break in the future?

**Comment by jennybc** [comment]
@olivroy something that puzzles me. Using `list()` versus `c()` for specifying line and column doesn't seem to affect the hyperlink. So I don't see how that can explain a change of behaviour.

``` r
> dput(cli::style_hyperlink(
+   "file","file://NEWS.md", list(line = 6, col = 2)
+ ))
structure("\033]8;line = 6:col = 2;file://NEWS.md\afile\033]8;;\a", class = c("cli_ansi_string", 
"ansi_string", "character"))
> dput(cli::style_hyperlink(
+   "file","file://NEWS.md", c(line = 6, col = 2)
+ ))
structure("\033]8;line = 6:col = 2;file://NEWS.md\afile\033]8;;\a", class = c("cli_ansi_string", 
"ansi_string", "character"))
```

**Reply by olivroy** [reply]
I don't know.. I mis-diagnosed the issue in this case. It seems that the syntax works for full paths, but not for relative paths... 

```r
# this works
cli::style_hyperlink(
  "file","file://C:/users/royo/documents/projects/infrastructure/NEWS.md", c(line = 6, col = 2)
 )
# this doesn't
cli::style_hyperlink(
  "file","file://NEWS.md", c(line = 6, col = 2)
 )
```

Both work in RStudio..

**Reply by jennybc** [reply, marked as answer]
I did a lot of research on file hyperlinks when working on #2413. I think the best outside authority to consult is RFC 8089 The "file" URI Scheme:

https://datatracker.ietf.org/doc/html/rfc8089

I don't really see any support for relative paths in such links. Leading up to now, I think it's fair to say that cli hyperlinks and RStudio support of said hyperlinks sort of co-evolved to perhaps handle nonstandard hyperlinks or to go beyond the standard. So this may not be an area where Positron implements 100% of what RStudio does. Positron will likely stick close to the standard.

My guess is that, if you are creating a hyperlink yourself, your code should look like this to have the best chance of working in multiple IDEs, across multiple OSes, over time:

```
cli::style_hyperlink(
  "file","file:///C:/Users/jenny/usethis/NEWS.md", c(line = 6, col = 2)
)
```

Key points:

* Use an absolute path
* Start with `file:///` (three slashes)
* Consistently use slash (`/`) as the path separator

``` r
cli::style_hyperlink(
  "file","file:///C:/users/royo/documents/projects/infrastructure/NEWS.md", c(line = 6, col = 2)
 )
```

**Reply by olivroy** [reply]
Thank you very much for taking the time!
